tests/xharness/Utilities/StringUtils.cs

```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Xharness.Utilities
+{
+	class StringUtils
+	{
+		static readonly char shellQuoteChar;
+		static readonly char [] mustQuoteCharacters = new char [] { ' ', '\'', ',', '$', '\\' };
+		static readonly char [] mustQuoteCharactersProcess = { ' ', '\\', '"', '\'' };
+
+		static StringUtils ()
+		{
+			PlatformID pid = Environment.OSVersion.Platform;
+			if ((int) pid != 128 && pid != PlatformID.Unix && pid != PlatformID.MacOSX)
+				shellQuoteChar = '"'; // Windows
+			else
+				shellQuoteChar = '\''; // !Windows
+		}
+
+		public static string FormatArguments (params string [] arguments)
+		{
+			return FormatArguments ((IList<string>) arguments);
+		}
+
+		public static string FormatArguments (IList<string> arguments)
+		{
+			return string.Join (" ", QuoteForProcess (arguments));
+		}
+
+		private static string [] QuoteForProcess (params string [] array)
+		{
+			if (array == null || array.Length == 0)
+				return array;
+
+			var rv = new string [array.Length];
+			for (var i = 0; i < array.Length; i++)
+				rv [i] = QuoteForProcess (array [i]);
+			return rv;
+		}
+
+		public static string Quote (string f)
+		{
+			if (string.IsNullOrEmpty (f))
+				return f ?? string.Empty;
+
+			if (f.IndexOfAny (mustQuoteCharacters) == -1)
+				return f;
+
+			var s = new StringBuilder ();
+
+			s.Append (shellQuoteChar);
+			foreach (var c in f) {
+				if (c == '\'' || c == '"' || c == '\\')
+					s.Append ('\\');
+
+				s.Append (c);
+			}
+			s.Append (shellQuoteChar);
+
+			return s.ToString ();
+		}
+
+		// Quote input according to how System.Diagnostics.Process needs it quoted.
+		private static string QuoteForProcess (string f)
+		{
+			if (string.IsNullOrEmpty (f))
+				return f ?? string.Empty;
+
+			if (f.IndexOfAny (mustQuoteCharactersProcess) == -1)
+				return f;
+
+			var s = new StringBuilder ();
+
+			s.Append ('"');
+			foreach (var c in f) {
+				if (c == '"') {
+					s.Append ('\\');
+					s.Append (c).Append (c);
```

Comment:
@rolfbjarne can you check this line? It was copied over from tools but through unit tests I discovered that this line does something weird with the output and I am not sure it is 100% correct.
```x"yz```
gets turned into
```x\"""yz```
(both then wrapped in "")
I am not sure this shouldn't have been appending `\` instead or something but I don't really understand the use case.

Reply:
Yes, that does look strange. That said, if unit tests like this work:

```csharp
using System.Diagnostics;
var p = new Process ();
p.StartInfo.RedirectStandardOutput = true;
p.StartInfo.UseShellExecute = false;
p.StartInfo.FileName = "/bin/echo";                                                                                                                                                                                                                                                                                             
var input = "\"""'"; // some complex input
p.StartInfo.Arguments = "-e " + StringUtils.QuoteForProcess (input);
p.Start ();
var output = p.StandardOutput.ReadToEnd ();
Assert.AreEqual (input, output, "echo");
```

then any changes should be good.